src/build_system.ml

```diff
@@ -1366,6 +1371,34 @@ end = struct
   let evaluate_action_and_dynamic_deps =
     Memo.exec evaluate_action_and_dynamic_deps_memo
 
+  let select_sandbox_mode
+        (config : Sandbox_config.t) ~loc ~sandboxing_preference =
+    match
+      List.find_map sandboxing_preference ~f:(fun preference ->
+        match (preference, config) with
+        | None, { none = true; _ } ->
+          Some None
+        | Some Sandbox_mode.Copy, { copy = true; _ } ->
+          Some (Some Sandbox_mode.Copy)
+        | Some Symlink, { symlink = true; copy; _ } ->
+          (if copy then
+             Some (Some (if Sys.win32 then Copy else Symlink))
+           else
+             Code_error.raise
+               "This rule requires sandboxing with symlinks, but that won't \
```

Comment:
It seems like we should use `User_error.raise ~loc` here. Otherwise it's not clear which rule we are talking about